File: packages/dart/lib/Seguranca/Ataques/AtaqueCusum.dart
import 'dart:math';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// Somas cumulativas (cusum), teste do NIST SP800-22. Converte os bits em
/// passos +1/-1 e observa o maior desvio da caminhada aleatória. Um viés
/// pequeno que o monobit quase não vê faz o desvio acumulado crescer.
/// Para uma sequência aleatória, max|S|/sqrt(n) fica tipicamente abaixo de 3.
class AtaqueCusum extends Ataque {
  final int _tamanho;
  final double _limiteZ;

  AtaqueCusum({int tamanho = 8192, double limiteZ = 4.0})
    : _tamanho = tamanho,
      _limiteZ = limiteZ;

  @override
  String nome() => 'Somas cumulativas (cusum)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final ks = alvo.gerarKeystreamBruto(
      alvo.chaveDeTeste(),
      randomBytes(alvo.tamanhoIv()),
      'enc',
      _tamanho,
    );

    var soma = 0;
    var maxAbs = 0;
    for (var i = 0; i < ks.length; i++) {
      final byte = ks[i];
      for (var b = 7; b >= 0; b--) {
        soma += ((byte >> b) & 1) == 1 ? 1 : -1;
        maxAbs = max(maxAbs, soma.abs());
      }
    }
    final n = ks.length * 8;
    final z = maxAbs / sqrt(n);
    final vulneravel = z > _limiteZ;

    return ResultadoAtaque(
      nome(),
      vulneravel,
      vulneravel ? Severidade.media : Severidade.info,
      'max|S|=$maxAbs sobre $n bits, '
          'max|S|/sqrt(n)=${z.toStringAsFixed(2)} '
          '(limite=${_limiteZ.toStringAsFixed(1)})',
      {'max_abs': maxAbs, 'z': z},
    );
  }
}

File: packages/rust/src/seguranca/ataques/ataque_cusum.rs
//! Somas cumulativas (cusum), teste do NIST SP800-22. Converte os bits em
//! passos +1/-1 e observa o maior desvio da caminhada aleatória. Um viés
//! pequeno que o monobit quase não vê faz o desvio acumulado crescer.
//! Para uma sequência aleatória, max|S|/sqrt(n) fica tipicamente abaixo de 3.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::random_bytes;

pub struct AtaqueCusum {
    pub tamanho: usize,
    pub limite_z: f64,
}

impl AtaqueCusum {
    pub fn new() -> Self {
        AtaqueCusum {
            tamanho: 8192,
            limite_z: 4.0,
        }
    }
}

impl Default for AtaqueCusum {
    fn default() -> Self {
        Self::new()
    }
}

impl Ataque for AtaqueCusum {
    fn nome(&self) -> String {
        "Somas cumulativas (cusum)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let ks = alvo.gerar_keystream_bruto(
            alvo.chave_de_teste().as_bytes(),
            &random_bytes(alvo.tamanho_iv()),
            "enc",
            self.tamanho,
        );

        let mut soma: i64 = 0;
        let mut max_abs: i64 = 0;
        for &byte in &ks {
            for b in (0..8).rev() {
                soma += if ((byte >> b) & 1) == 1 { 1 } else { -1 };
                max_abs = max_abs.max(soma.abs());
            }
        }
        let n = ks.len() * 8;
        let z = max_abs as f64 / (n as f64).sqrt();
        let vulneravel = z > self.limite_z;

        Ok(ResultadoAtaque::new(
            &self.nome(),
            vulneravel,
            if vulneravel {
                Severidade::Media
            } else {
                Severidade::Info
            },
            &format!(
                "max|S|={} sobre {} bits, max|S|/sqrt(n)={:.2} (limite={:.1})",
                max_abs, n, z, self.limite_z
            ),
        ))
    }
}
